lib/index.js

```diff
@@ -1,44 +1,124 @@
-var Winston = require('winston');
+const Confit = require('confit')
+const express = require('express')
+const Middleware = require('./middleware')
+const Path = require('path')
+const Routes = require('./routes')
+const URL = require('url')
+const Utils = require('./utils')
+const Winston = require('winston')
+const medUtils = require('openhim-mediator-utils')
+const mediatorConfig = require('../config/mediator')
+const Cache = require('./cache')
 
-var app = Express();
+var server
 
-// Load the config
-Confit(Path.join(__dirname, '..', 'config')).create(function(err, config) {
-  if (err) {
-    throw err;
-  }
+function setupApp (config, appConfig) {
+  const app = express()
 
   // Make the config accessible to all routes
-  Object.defineProperty(app.locals, 'config', {value: config});
-
-  Winston.clear();
-  Winston.add(Winston.transports.Console, {timestamp: true, level: config.get('logging:level')});
+  Object.defineProperty(app.locals, 'config', { value: config })
 
-  var url = URL.resolve(config.get('dhis:url'), config.get('dhis:path'));
-  // Populate the cache on startup
-  Utils.populateCache(url);
-  // Repopulate the cache based on the cron pattern
-  new CronJob(config.get('cron:pattern'), Utils.populateCache.bind(null, url), true, config.get('cron:timezone'));
+  Winston.clear()
+  Winston.add(Winston.transports.Console, { timestamp: true, level: appConfig.get('logging:level') })
 
   // Set up the routes
-  app.get('/heartbeat', Routes.uptime);
-  app.use(Middleware.stats(config.get('statsd')));
-  app.get(config.get('dhis:path'), Routes.facilityCodeLookup);
+  app.get('/heartbeat', Routes.uptime)
+  app.use(Middleware.stats(appConfig.get('statsd')))
+
+  config.routes.forEach((route) => {
+    var url = URL.resolve(route.dhisUrl, route.dhisPath)
+
+    // Populate cache on startup
+    Utils.populateCache(url)
+    Utils.addCronJob(route.cronPattern, route.cronTimezone, url)
+    Utils.addRoute(app, route.dhisPath)
+  })
+  return app
+}
+exports.setupApp = setupApp
+
+// Load the config
+function start (callback) {
+  setupConfig((err, appConfig) => {
+    if (err) {
+      callback(err)
+    }
+
+    let apiConf = appConfig.get('openhim')
+    let port = appConfig.get('server:port')
+
+    if (apiConf.api.trustSelfSigned) {
+      process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+    }
+
+    if (apiConf.register) {
+      medUtils.registerMediator(apiConf.api, mediatorConfig, (err) => {
+        if (err) {
+          Winston.error('Failed to register this mediator, check your config')
+          Winston.error(err.stack)
+          process.exit(1)
+        }
+        apiConf.api.urn = mediatorConfig.urn
+        medUtils.fetchConfig(apiConf.api, (err, newConfig) => {
+          Winston.info('Received initial config:')
+          Winston.info(JSON.stringify(newConfig))
+          let config = newConfig
+          if (err) {
+            Winston.error('Failed to fetch initial config')
+            Winston.error(err.stack)
```

Comment:
The error should be passed as metadata to the log statement above. Winston automatically handles printing the stack.